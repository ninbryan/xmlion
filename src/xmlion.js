const {isArray} = Array;
const clone = (array) => array.slice(0);
const join = (array) => array.join('');
const map = (fn, array) => array.map(fn);
const filter = (fn, array) => array.filter(fn);
const concat = (array, ...arrays) => array.concat(...arrays);
const toStringCub = (c) => c == null ? '' : c.ROARS ? c._buildElementString() : c;
const mapCubsToString = (cubs) => map(toStringCub, cubs);

export class Lion {
  constructor (tagName, attributes, cubs) {
    this.tagName = tagName;
    this.attr = attributes || {};
    this.cubs = cubs || [];
  }

  get ROARS () {
    return true;
  }

  set cubs (cubs) {
    if (isArray(cubs)) {
      this._cubs = cubs;
    } else {
      throw new TypeError('Lion wants cub[]');
    }
  }

  get cubs () {
    return clone(this._cubs);
  }

  set attributes (attr) {
    this.attr = attr;
  }

  get attributes () {
    return this.attr;
  }

  setCubs (cubs) {
    this.cubs = cubs;
    return this;
  }

  getCubs () {
    return this.cubs;
  }

  setAttributes (attr) {
    this.attributes = attr;
    return this;
  }

  getAttributes () {
    return this.attributes;
  }

  handleBuildAttribute (key, value) {
    if (value === null || value === undefined) {
      return key;
    }
    return `${key}="${value}"`;
  }

  _buildAttributeString () {
    const attr = this.attr;
    const keys = Object.keys(attr).sort();
    if (keys.length > 0) {
      return join(map((key) => ` ${this.handleBuildAttribute(key, attr[key])}`, keys));
    }
    return '';
  }

  _buildContentString () {
    return join(mapCubsToString(this._cubs));
  }

  _buildElementString () {
    const me = this;
    const tagName = me.tagName;
    const attributes = me._buildAttributeString();

    if (me.hasCubs()) {
      const content = me._buildContentString();

      return `<${tagName}${attributes}>${content}</${tagName}>`;
    } else {
      return `<${tagName}${attributes}/>`;
    }
  }

  addAttribute (name, value) {
    const lion = this;
    lion.attr[name] = value;
    return lion;
  }

  removeAttribute (name) {
    const lion = this;
    if (lion.attr[name]) {
      delete lion.attr[name];
    }
    return lion;
  }

  addAttributes (attr) {
    const lion = this;
    const names = Object.keys(attr);
    map((name) => lion.addAttribute(name, attr[name]), names);
    return lion;
  }

  addCub (cub) {
    const lion = this;
    lion._cubs.push(cub);
    return lion;
  }

  removeCub (cub) {
    let lion = this;
    let index = lion._cubs.indexOf(cub);
    if (index >= 0) {
      lion.cubs = filter((c, i) => (i !== index), lion.cubs);
    }
    return lion;
  }

  addCubs (cubs) {
    if (isArray(cubs)) {
      this.cubs = concat(this.cubs, cubs);
    }
    return this;
  }

  removeAllCubs () {
    this._cubs = [];
    return this;
  }

  hasCubs () {
    return this._cubs && this._cubs.length > 0;
  }

  value () {
    return this._buildElementString();
  }
}

export const xmlion = (tagName, attributes, cubs) => new Lion(tagName, attributes, cubs);
