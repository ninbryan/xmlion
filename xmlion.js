(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.xmlion = f()}})(function(){var define,module,exports;module={exports:(exports={})};
'use strict';

var ROARS = true;

function Lion(tagName, attributes, cubs) {
    if (!this || !this.ROARS) {
        return new Lion(tagName, attributes, cubs);
    }

    this.tagName = tagName;
    this.attr = attributes || {};
    this.cubs = cubs || [];
}

Lion.prototype.ROARS = ROARS;

Lion.prototype._buildAttributeString = function () {
    var _this = this;

    var attributes = '';
    var keys = Object.keys(this.attr);
    if (keys.length > 0) {
        keys.forEach(function (key) {
            attributes += ' ' + key + '="' + _this.attr[key] + '"';
        });
    }
    return attributes;
};

Lion.prototype._buildContentString = function () {
    return this.cubs.map(function (cub) {
        if (cub == null) {
            return '';
        }
        if (cub.ROARS) {
            return cub._buildElementString();
        }

        return cub.toString();
    }).join('');
};

Lion.prototype._buildElementString = function () {
    var me = this;
    var tagName = me.tagName;
    var attributes = me._buildAttributeString();

    if (me.hasCubs()) {
        var content = me._buildContentString();

        return '<' + tagName + attributes + '>' + content + '</' + tagName + '>';
    } else {
        return '<' + tagName + attributes + '/>';
    }
};

Lion.prototype.addAttribute = function (name, value) {
    this.attr[name] = value;
    return this;
};

Lion.prototype.removeAttribute = function (name) {
    if (this.attr[name]) {
        delete this.attr[name];
    }
    return this;
};

Lion.prototype.addAttributes = function (attr) {
    var lion = this;
    var names = Object.keys(attr);
    var len = names.length;
    for (var x = 0; x < len; x++) {
        var name = names[x];
        lion.addAttribute(name, attr[name]);
    }
    return lion;
};

Lion.prototype.hasCubs = function () {
    return this.cubs && this.cubs.length > 0;
};

Lion.prototype.value = function () {
    return this._buildElementString();
};

module.exports = Lion;
return module.exports;});
