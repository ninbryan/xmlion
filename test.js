import xmlion from './src/index';
import test from 'tape';

test('empty elements', (assert) => {
    let actual = xmlion('lion').value();
    
    assert.equal(actual, '<lion/>', 'xmlion must make empty elements');
    
    assert.end();
});

test('attributed elements', (assert) => {
    let actual = xmlion('lion', {attr:'test'}).value();
    
    assert.equal(actual, '<lion attr="test"/>', 'xmlion must make attributed elements');
    
    assert.end();
});

test('attributed elements with content', (assert) => {
    let actual = xmlion('lion', {attr:'test'}, ['test']).value();
    
    assert.equal(actual, '<lion attr="test">test</lion>', 'xmlion must make attributed elements with content');
    
    assert.end();
});

test('attributed elements with two cubs', (assert) => {
    let actual = xmlion('lion', {attr:'test'}, ['test',xmlion('lion', {attr:'test'}, ['test'])]).value();
    
    assert.equal(actual, '<lion attr="test">test<lion attr="test">test</lion></lion>', 'xmlion must make complex attributed elements with content');
    
    assert.end();
});

test('attributed elements with empty values', (assert) => {
    let actual = xmlion('lion', {attr:'test'}, [null,undefined,NaN,0]).value();
    
    assert.equal(actual, '<lion attr="test">NaN0</lion>', 'xmlion must make complex attributed elements with content');
    
    assert.end();
});


const lion = xmlion('lion', {}, []);
    
test('add attribute', (assert) => {
    
    lion.addAttribute('attr','add');
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr="add"/>', 'xmlion must add attribute');
    
    assert.end();
});

test('remove attribute', (assert) => {
    
    lion.removeAttribute('attr');
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion/>', 'xmlion must remove attribute');
    
    assert.end();
});

test('add many attributes', (assert) => {
    
    lion.addAttributes({
        'attr1':'yup',
        'attr2':'yo'
    });
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr1="yup" attr2="yo"/>', 'xmlion must add multiple attributes');
    
    assert.end();
});

const cub = xmlion('cub');

test('add cub', (assert) => {
    
    lion.addCub(cub);
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr1="yup" attr2="yo"><cub/></lion>', 'xmlion must add a cub');
    
    assert.end();
});

test('remove cub', (assert) => {
    
    lion.removeCub(cub);
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr1="yup" attr2="yo"/>', 'xmlion must remove a cub');
    
    assert.end();
});

test('add cubs', (assert) => {
    
    lion.addCubs([cub, xmlion('cub', null, ['meow'])]);
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr1="yup" attr2="yo"><cub/><cub>meow</cub></lion>', 'xmlion must add many cubs');
    
    assert.end();
});

test('remove all cubs', (assert) => {
    
    lion.removeAllCubs();
    
    let actual = lion.value();
    
    assert.equal(actual, '<lion attr1="yup" attr2="yo"/>', 'xmlion must add a cub');
    
    assert.end();
});