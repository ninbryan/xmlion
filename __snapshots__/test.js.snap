// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version 1`] = `"1.0.0"`;
